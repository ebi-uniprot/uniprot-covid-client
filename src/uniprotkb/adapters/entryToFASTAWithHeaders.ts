import { formatFASTA } from 'franklin-sites';

import { UniProtkbAPIModel, EntryType } from './uniProtkbConverter';
import { APISequenceData } from '../../tools/blast/types/apiSequenceData';

type Subset = { start: number; end: number };

type Modifications = { subsets: Subset[] }; // keep open for variant for example?

// build a "nicely"-formatted FASTA string
// See https://www.uniprot.org/help/fasta-headers for current headers
const entryToFASTAWithHeaders = (
  entry: UniProtkbAPIModel | APISequenceData,
  modifications?: Modifications
): string => {
  let sequence = entry.sequence.value || '';

  const subsets = [];
  // if any change is required on the sequence, do it here
  if (modifications) {
    if (modifications.subsets) {
      let subsetSequence = '';
      for (const { start, end } of modifications.subsets) {
        subsetSequence += sequence.slice(start - 1, end);
        subsets.push(`${start}-${end}`);
      }
      sequence = subsetSequence;
    }
  }

  try {
    let db;
    switch (entry.entryType) {
      case EntryType.REVIEWED:
        db = 'sp';
        break;
      case EntryType.UNREVIEWED:
        db = 'tr';
        break;
      default:
        db = '??';
    }
    let optionalProteinName = '';
    if (entry?.proteinDescription?.recommendedName?.fullName) {
      optionalProteinName = `${entry.proteinDescription.recommendedName.fullName.value} `;
    }
    let optionalOS = '';
    if (entry?.organism?.scientificName) {
      optionalOS = `OS=${entry.organism.scientificName} `;
    }
    let optionalOX = '';
    if (entry?.organism?.taxonId) {
      optionalOX = `OX=${entry.organism.taxonId} `;
    }
    let optionalGN = '';
    if (entry.genes?.[0]?.geneName?.value) {
      optionalGN = `GN=${entry.genes[0].geneName.value} `;
    }
    const pe = entry.proteinExistence[0];
    let optionalSV = '';
    if (entry?.entryAudit?.sequenceVersion) {
      optionalSV = `SV=${entry.entryAudit.sequenceVersion}`;
    }
    let optionalSubset = '';
    if (subsets.length) {
      optionalSubset = `|${subsets.join(',')}`;
    }
    sequence = `>${db}|${entry.primaryAccession}|${entry.uniProtkbId}${optionalSubset} ${optionalProteinName}${optionalOS}${optionalOX}${optionalGN}PE=${pe} ${optionalSV}\n${sequence}`;
  } catch {
    // temporary catch for uniParc entry
    if ((entry as any).uniParcId) {
      sequence = `>${(entry as any).uniParcId}\n${sequence}`;
    } else {
      // empty header 🤷🏽‍♂️
      sequence = `>\n${sequence}`;
    }
  }
  return formatFASTA(sequence);
};

export default entryToFASTAWithHeaders;
