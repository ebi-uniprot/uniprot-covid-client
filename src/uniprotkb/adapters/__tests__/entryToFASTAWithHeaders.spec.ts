import entryToFASTAWithHeaders from '../entryToFASTAWithHeaders';

import modelData from '../../__mocks__/entryModelData.json';

describe('entryToFASTAWithHeaders', () => {
  it('should output a FASTA string with UniProt-style headers', () => {
    const fasta = entryToFASTAWithHeaders(modelData);
    expect(fasta).toMatchSnapshot();
  });

  it('should output a FASTA with subset', () => {
    const fasta = entryToFASTAWithHeaders(modelData, {
      subsets: [
        { start: 1, end: 1 },
        { start: 3, end: 5 },
      ],
    });
    expect(fasta).toMatchSnapshot();
  });
});
