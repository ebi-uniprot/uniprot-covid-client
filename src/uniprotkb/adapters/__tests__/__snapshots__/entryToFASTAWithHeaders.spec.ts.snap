// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entryToFASTAWithHeaders should output a FASTA string with UniProt-style headers 1`] = `
">sp|P21802|uniprot_id rec full Name OS=scientific name OX=9606 GN=some Gene PE=1 SV=5
SAPSQDFMRF"
`;

exports[`entryToFASTAWithHeaders should output a FASTA with subset 1`] = `
">sp|P21802|uniprot_id|1-1,3-5 rec full Name OS=scientific name OX=9606 GN=some Gene PE=1 SV=5
SPSQ"
`;
